File: sdk-dotnet/PathPrefixes.cs
namespace Tollara;

/// <summary>ECS vs Docker path prefixes for hosted Tollara API origins (parity with sdk-js).</summary>
public static class PathPrefixes
{
    public const string EcsCorePathPrefix = "/core/api/v1";
    public const string EcsGatewayPathPrefix = "/gateway/api/v1";
    public const string EcsUsagePathPrefix = "/usage/api/v1";

    public static bool IsHostedTollaraApiOrigin(string origin)
    {
        if (string.IsNullOrWhiteSpace(origin)) return false;
        if (!Uri.TryCreate(origin.Trim(), UriKind.Absolute, out var uri) || uri.Host is null)
            return false;
        var host = uri.Host.ToLowerInvariant();
        if (host == "api.tollara.ai" || host.EndsWith(".api.tollara.ai", StringComparison.Ordinal))
            return true;
        return host == "api.ppe.tollara.ai" || host.EndsWith(".api.ppe.tollara.ai", StringComparison.Ordinal);
    }

    public static string ResolveGatewayPathPrefix(string? baseUrl, string? overridePrefix = null)
    {
        if (!string.IsNullOrWhiteSpace(overridePrefix)) return overridePrefix.Trim();
        var origin = ResolveOrigin(baseUrl);
        return IsHostedTollaraApiOrigin(origin) ? EcsGatewayPathPrefix : TollaraClient.DefaultGatewayPathPrefix;
    }

    public static string ResolveCorePathPrefix(string? baseUrl, string? overridePrefix = null)
    {
        if (!string.IsNullOrWhiteSpace(overridePrefix)) return overridePrefix.Trim();
        var origin = ResolveOrigin(baseUrl);
        return IsHostedTollaraApiOrigin(origin) ? EcsCorePathPrefix : TollaraClient.DefaultCorePathPrefix;
    }

    public static string ResolveUsagePathPrefix(string? baseUrl, string? overridePrefix = null)
    {
        if (!string.IsNullOrWhiteSpace(overridePrefix)) return overridePrefix.Trim();
        var origin = ResolveOrigin(baseUrl);
        return IsHostedTollaraApiOrigin(origin) ? EcsUsagePathPrefix : TollaraClient.DefaultUsagePathPrefix;
    }

    private static string ResolveOrigin(string? baseUrl)
    {
        if (!string.IsNullOrWhiteSpace(baseUrl))
        {
            var t = baseUrl.Trim();
            while (t.EndsWith('/')) t = t[..^1];
            return t;
        }
        return TollaraClient.DefaultApiUrl;
    }
}

File: sdk-php/src/PathPrefixes.php
<?php

declare(strict_types=1);

namespace Tollara\ServiceSdk;

final class PathPrefixes
{
    public const ECS_CORE_PATH_PREFIX = '/core/api/v1';
    public const ECS_GATEWAY_PATH_PREFIX = '/gateway/api/v1';
    public const ECS_USAGE_PATH_PREFIX = '/usage/api/v1';

    public static function isHostedTollaraApiOrigin(string $origin): bool
    {
        $host = parse_url(trim($origin), PHP_URL_HOST);
        if (!is_string($host) || $host === '') {
            return false;
        }
        $host = strtolower($host);
        if ($host === 'api.tollara.ai' || str_ends_with($host, '.api.tollara.ai')) {
            return true;
        }
        return $host === 'api.ppe.tollara.ai' || str_ends_with($host, '.api.ppe.tollara.ai');
    }

    public static function resolveGatewayPathPrefix(?string $baseUrl, ?string $override = null): string
    {
        if ($override !== null && trim($override) !== '') {
            return trim($override);
        }
        $origin = self::resolveOrigin($baseUrl);
        return self::isHostedTollaraApiOrigin($origin)
            ? self::ECS_GATEWAY_PATH_PREFIX
            : TollaraClient::DEFAULT_GATEWAY_PATH_PREFIX;
    }

    public static function resolveCorePathPrefix(?string $baseUrl, ?string $override = null): string
    {
        if ($override !== null && trim($override) !== '') {
            return trim($override);
        }
        $origin = self::resolveOrigin($baseUrl);
        return self::isHostedTollaraApiOrigin($origin)
            ? self::ECS_CORE_PATH_PREFIX
            : TollaraClient::DEFAULT_CORE_PATH_PREFIX;
    }

    public static function resolveUsagePathPrefix(?string $baseUrl, ?string $override = null): string
    {
        if ($override !== null && trim($override) !== '') {
            return trim($override);
        }
        $origin = self::resolveOrigin($baseUrl);
        return self::isHostedTollaraApiOrigin($origin)
            ? self::ECS_USAGE_PATH_PREFIX
            : TollaraClient::DEFAULT_USAGE_PATH_PREFIX;
    }

    private static function resolveOrigin(?string $baseUrl): string
    {
        if ($baseUrl !== null && trim($baseUrl) !== '') {
            return rtrim(trim($baseUrl), '/');
        }
        return TollaraClient::DEFAULT_API_URL;
    }
}
